File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function smallestFromLeaf(root: TreeNode | null): string {
    const preorder = (root: TreeNode| null, strArr: string[]|null, prev: string="") :void =>{
        if(root === null)
            return;
            
        const str: string = prev + String.fromCharCode(root.val + 97);
        if(root.left === null && root.right === null){
            let t = str.split("").reverse().join("");
            strArr.push(t);
        }
        preorder(root.left, strArr, str);
        preorder(root.right, strArr, str);
    }  

    const leafArr: string[] = [];
    preorder(root, leafArr, "");
    const ans: string = leafArr.reduce((next, prev) => next < prev ? next : prev);
    return ans;
};

File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn smallest_from_leaf(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        let mut smallest_string = "~".to_owned();
        let mut stack = vec![(root.unwrap(), vec![])];
        while !stack.is_empty() {
            let (node, mut letters) = stack.pop().unwrap(); // safety in case stack is non-empty
            let node_ref = node.borrow();
            letters.insert(0, node_ref.val as u8 + b'a');
            if node_ref.left.is_none() && node_ref.right.is_none() { // leaf node
                let string = String::from_utf8(letters).unwrap();
                smallest_string = smallest_string.min(string);
            } else {
                if let Some(left) = node_ref.left.clone() {
                    stack.push((left, letters.clone()));
                }
                if let Some(right) = node_ref.right.clone() {
                    stack.push((right, letters));
                }
            }
        }
        smallest_string
    }
}